algosdk/future/atomic_transaction_composer.py

```diff
@@ -0,0 +1,512 @@
+        """
+        Send the transaction group to the network and wait until it's committed
+        to a block. An error will be thrown if submission or execution fails.
+        The composer's status must be SUBMITTED or lower before calling this method,
+        since execution is only allowed once. If submission is successful,
+        this composer's status will update to SUBMITTED.
+        If the execution is also successful, this composer's status will update to COMMITTED.
+
+        Note: a group can only be submitted again if it fails.
+
+        Returns:
+            AtomicTransactionResponse: Object with confirmed round for this transaction,
+                a list of txIDs of the submitted transactions, and an array of
+                results for each method call transaction in this group. If a
+                method has no return value (void), then the method results array
+                will contain None for that method's return value.
+        """
+        if self.status > AtomicTransactionComposerStatus.SUBMITTED:
+            raise error.AtomicTransactionComposerError(
+                "AtomicTransactionComposerStatus must be submitted or lower to execute a group"
+            )
+
+        self.submit(client)
+        transaction.wait_for_confirmation(client, self.tx_ids[0], wait_rounds)
+        self.status = AtomicTransactionComposerStatus.COMMITTED
+
+        confirmed_round = -1
+        method_results = []
+
+        for i, tx_id in enumerate(self.tx_ids):
+            if i not in self.method_dict:
+                continue
+            # Return is void
+            if self.method_dict[i].returns is None:
+                method_results.append(
+                    ABIResult(
+                        tx_id=tx_id,
+                        raw_value=None,
+                        return_value=None,
+                        decode_error=None,
+                    )
+                )
+                continue
+
+            # Parse log for ABI method return value
+            resp = client.pending_transaction_info(tx_id)
+            confirmed_round = resp["confirmed-round"]
+            logs = resp["logs"] if "logs" in resp else None
+            raw_value = None
+            return_value = None
+            decode_error = None
+            # Look for the last returned value in the log
+            for result in reversed(logs):
+                # Check that the first four bytes is the hash of "return"
+                result_bytes = base64.b64decode(result)
+                if result_bytes[:4] != bytes.fromhex(ABI_RETURN_HASH):
+                    continue
+                try:
+                    raw_value = result_bytes[4:]
+                    return_value = self.method_dict[i].returns.type.decode(
+                        raw_value
+                    )
+                    decode_error = None
+                except Exception as e:
+                    raw_value = None
```

Comment:
No need to erase `raw_value` if an exception happens, since if the type decoding fails, it's still worthwhile to return the raw value